### PR TITLE
[server] Report Stripe client metrics

### PR DESCRIPTION
**Pull request by easyCZ**
## Description

Reports prometheus metrics on Stripe interaction.

## Related Issue(s)

* Part of https://github.com/gitpod-io/gitpod/issues/13157
* Needed to get data for https://github.com/gitpod-io/gitpod/issues/13177 when migrating

## How to test

Creating a UBP subscription works

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-payment
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

### CODE REVIEW

File: components/server/ee/src/user/stripe-service.ts [hunk 1]
```diff
@@ -149,11 +180,27 @@ export class StripeService {
...
+                customer: customer.id,
+                items: [{ price: priceId }],
+                automatic_tax: { enabled: isAutomaticTaxSupported },
+                billing_cycle_anchor: Math.round(startOfNextMonth.getTime() / 1000),
+            });
         });
     }
 }
+
+async function reportStripeOutcome<T>(op: string, f: () => T) {
```
Comment (geropl) [inline components/server/ee/src/user/stripe-service.ts:195]:
```suggestion
async function reportStripeOutcome<T>(op: string, f: () => Promise<T>): T {
```

Comment (easyCZ) [reply components/server/ee/src/user/stripe-service.ts:195]: What's the difference? T can be either, no?

Comment (easyCZ) [reply components/server/ee/src/user/stripe-service.ts:195]: I guess it forces the return to be a promise

Comment (geropl) [reply components/server/ee/src/user/stripe-service.ts]: :+1: It avoids copy'n'paste errors where you insert an `await`, or forget what you're supposed to return.

Comment (easyCZ) [reply components/server/ee/src/user/stripe-service.ts]: That makes sense, thanks for the tip!

File: components/server/ee/src/user/stripe-service.ts [hunk 2]
```diff
@@ -149,11 +180,27 @@ export class StripeService {
...
     }
 }
+
+async function reportStripeOutcome<T>(op: string, f: () => Promise<T>) {
+    const timer = stripeClientRequestsCompletedDurationSeconds.startTimer();
+    let outcome = "ok";
+    try {
+        return await f();
+    } catch (e) {
+        outcome = e.type;
```
Comment (andrew-farries) [inline components/server/ee/src/user/stripe-service.ts:201]: e has `any` type here. How do we know that it has a `type` field?

**Review by andrew-farries — APPROVED** [review]
/hold for question

File: components/server/ee/src/user/stripe-service.ts (hunk 2 above)
Comment (geropl) [reply components/server/ee/src/user/stripe-service.ts:201]: We don't. But you're right, outcome should be `string | undefined`, and ` observeStripeClientRequestsCompleted` might need minimal extension to translate `undefined` to `"undefined"` (or whatever we want to call it).

Comment (easyCZ) [reply components/server/ee/src/user/stripe-service.ts:201]: Interesting, I'd have hoped TS has my back here but looks like no. 

I mostly followed https://stripe.com/docs/api/errors/handling?lang=node

Will add a guard

Comment (easyCZ) [reply components/server/ee/src/user/stripe-service.ts]: Added.

**Review by geropl — APPROVED** [review]
Tested and worked. :ballot_box_with_check:

**Pull request merged**